File: tfa/tfa-core/src/main/java/tfa/model/Baseline.java
package tfa.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

/**
 * The consensus (what "normal" looks like) for one {@link FlowCluster}, computed
 * over collapsed call-site sequences. The population is the baseline — no golden
 * path is authored.
 */
public final class Baseline {

    /** A distinct sequence shape and the share of episodes matching it exactly. */
    public record SequenceShare(List<String> sequence, long count, double share) {}

    /** A call site observed at a modal position, with how often and its share. */
    public record PositionOption(String callSite, long count, double share) {}

    /** Median and p95 elapsed time for one collapsed transition A -> B. */
    public record TransitionTiming(String from, String to, long count,
                                   double medianMillis, double p95Millis) {}

    private final String clusterSignature;
    private final int episodesUsed;
    private final List<String> modalSequence;
    private final double modalShare;
    private final long modalCount;
    private final List<SequenceShare> alternatives;
    private final List<List<PositionOption>> positional;       // aligned to modalSequence
    private final Map<String, Map<String, Long>> transitionCounts;
    private final List<TransitionTiming> transitionTimings;

    public Baseline(String clusterSignature, int episodesUsed, List<String> modalSequence,
                    double modalShare, long modalCount, List<SequenceShare> alternatives,
                    List<List<PositionOption>> positional,
                    Map<String, Map<String, Long>> transitionCounts,
                    List<TransitionTiming> transitionTimings) {
        this.clusterSignature = clusterSignature;
        this.episodesUsed = episodesUsed;
        this.modalSequence = List.copyOf(modalSequence);
        this.modalShare = modalShare;
        this.modalCount = modalCount;
        this.alternatives = List.copyOf(alternatives);
        this.positional = List.copyOf(positional);
        this.transitionCounts = Map.copyOf(transitionCounts);
        this.transitionTimings = List.copyOf(transitionTimings);
    }

    public String clusterSignature()          { return clusterSignature; }
    public int episodesUsed()                 { return episodesUsed; }
    public List<String> modalSequence()       { return modalSequence; }
    public double modalShare()                { return modalShare; }
    public long modalCount()                  { return modalCount; }
    public List<SequenceShare> alternatives() { return alternatives; }

    /** Terminal call site of the modal sequence, or null if the modal sequence is empty. */
    public String modalTerminal() {
        return modalSequence.isEmpty() ? null : modalSequence.get(modalSequence.size() - 1);
    }

    /** Options observed at position {@code i} of the modal sequence, most frequent first. */
    public List<PositionOption> positionOptions(int i) {
        return (i < 0 || i >= positional.size()) ? List.of() : positional.get(i);
    }

    /** The majority call site expected at position {@code i}, or null if out of range. */
    public PositionOption expectedAt(int i) {
        List<PositionOption> opts = positionOptions(i);
        return opts.isEmpty() ? null : opts.get(0);
    }

    /** P(B follows A) within this cluster, over collapsed transitions. */
    public double transitionProbability(String from, String to) {
        Map<String, Long> outgoing = transitionCounts.get(from);
        if (outgoing == null || outgoing.isEmpty()) {
            return 0.0;
        }
        long total = outgoing.values().stream().mapToLong(Long::longValue).sum();
        return total == 0 ? 0.0 : (double) outgoing.getOrDefault(to, 0L) / total;
    }

    public List<TransitionTiming> transitionTimings() { return transitionTimings; }

    /** The {@code n} slowest transitions by p95. */
    public List<TransitionTiming> slowestByP95(int n) {
        List<TransitionTiming> sorted = new ArrayList<>(transitionTimings);
        sorted.sort(Comparator.comparingDouble(TransitionTiming::p95Millis).reversed());
        return sorted.subList(0, Math.min(n, sorted.size()));
    }

    /** Timing for a specific transition, or null if never observed with valid timestamps. */
    public TransitionTiming timingFor(String from, String to) {
        for (TransitionTiming t : transitionTimings) {
            if (t.from().equals(from) && t.to().equals(to)) {
                return t;
            }
        }
        return null;
    }
}

File: pytfa/tfa/baseline.py
"""Consensus baseline per cluster, over collapsed sequences. Port of
`tfa.baseline`. Deterministic (ties broken lexicographically)."""
from __future__ import annotations

import math
from typing import Optional

from .config import BaselineConfig
from .ingest import epoch_millis
from .model import (Baseline, Episode, FlowCluster, PositionOption, SequenceShare,
                    TransitionTiming)


def percentile(sorted_asc: list[int], p: float) -> float:
    if not sorted_asc:
        return 0.0
    n = len(sorted_asc)
    rank = math.ceil(p / 100.0 * n)
    rank = max(1, min(rank, n))
    return float(sorted_asc[rank - 1])


def build_baseline(cluster: FlowCluster, config: BaselineConfig) -> Optional[Baseline]:
    used = [e for e in cluster.episodes
            if config.in_baseline_window(e.start) and e.collapsed_sequence()]
    if not used:
        return None

    # 1. modal sequence + alternatives
    seq_counts: dict[tuple[str, ...], int] = {}
    for e in used:
        key = tuple(e.collapsed_sequence())
        seq_counts[key] = seq_counts.get(key, 0) + 1
    ranked = sorted(seq_counts.items(), key=lambda kv: (-kv[1], "".join(kv[0])))

    modal_sequence = list(ranked[0][0])
    modal_count = ranked[0][1]
    modal_share = modal_count / len(used)

    alternatives: list[SequenceShare] = []
    for seq, c in ranked[1:]:
        if len(alternatives) >= config.alternatives_to_report:
            break
        alternatives.append(SequenceShare(seq, c, c / len(used)))

    # 2. positional distribution
    positional: list[list[PositionOption]] = []
    for pos in range(len(modal_sequence)):
        at_pos: dict[str, int] = {}
        reached = 0
        for e in used:
            seq = e.collapsed_sequence()
            if pos < len(seq):
                reached += 1
                at_pos[seq[pos]] = at_pos.get(seq[pos], 0) + 1
        opts = [PositionOption(cs, n, (n / reached) if reached else 0.0)
                for cs, n in at_pos.items()]
        opts.sort(key=lambda o: (-o.count, o.call_site))
        positional.append(opts)

    # 3. transition counts + 4. timing samples
    transition_counts: dict[str, dict[str, int]] = {}
    timing_samples: dict[tuple[str, str], list[int]] = {}
    for e in used:
        runs = e.collapsed_runs()
        for i in range(len(runs) - 1):
            a, b = runs[i], runs[i + 1]
            transition_counts.setdefault(a.call_site, {})
            transition_counts[a.call_site][b.call_site] = \
                transition_counts[a.call_site].get(b.call_site, 0) + 1
            if a.first_timestamp is not None and b.first_timestamp is not None:
                ms = max(0, epoch_millis(b.first_timestamp) - epoch_millis(a.first_timestamp))
                timing_samples.setdefault((a.call_site, b.call_site), []).append(ms)

    timings: list[TransitionTiming] = []
    for (frm, to), samples in timing_samples.items():
        samples.sort()
        timings.append(TransitionTiming(frm, to, len(samples),
                                        percentile(samples, 50), percentile(samples, 95)))
    timings.sort(key=lambda t: (t.frm, t.to))

    return Baseline(cluster.signature, len(used), modal_sequence, modal_share,
                    modal_count, alternatives, positional, transition_counts, timings)
